File: AffineTransformationEx/Scala/src/main/scala/Affiex.scala
import org.opencv.highgui.Highgui
import visual.Imshow
import org.opencv.imgproc.Imgproc
import org.opencv.core.Mat
import org.opencv.core.CvType
import org.opencv.core.Size
import org.kohsuke.args4j.Option
import scala.collection.JavaConverters._
import org.kohsuke.args4j.CmdLineParser
import org.opencv.core.Core

/**
 * @author Depeng Liang
 */
object Affiex {
  
  nu.pattern.OpenCV.loadShared()
  
  def main(args: Array[String]): Unit = {
    val opt = new Argments()
    val parser: CmdLineParser = new CmdLineParser(opt)
    try {
      parser.parseArgument(args.toList.asJava)
      val img = Highgui.imread(opt.image)
      
      val oriShow = new Imshow("original image")
      oriShow.showImage(img)
      
      val bbox = opt.bbox.split(",").map(_.toInt)
      val outWh = opt.outWH.split(",").map(_.toInt)

      // crop mat, mapping the center of bbox to the center of croped image after affine transformation
      val cropMat = new Mat(3, 3, CvType.CV_32F)
      var tmpArr = Array[Float](
          1, 0, 0,
          0, 1, 0,
          0, 0, 1
      )
      cropMat.put(0, 0, tmpArr)

      val bboxCenterX = bbox(0) + bbox(2) / 2;
      val bboxCenterY = bbox(1) + bbox(3) / 2;

      val srcPt = Array(bboxCenterX, bboxCenterY) // bbox center
      val dstPt = Array(outWh(0) / 2, outWh(1) / 2) // croped image center

      val scaleX = outWh(0).toFloat / bbox(2)
      val scaleY = outWh(1).toFloat / bbox(3)
  
      tmpArr(0) = scaleX
      tmpArr(4) = scaleY
      
      tmpArr(2) = dstPt(0) - tmpArr(0) * srcPt(0)
      tmpArr(5) = dstPt(1) - tmpArr(4) * srcPt(1)
      cropMat.put(0, 0, tmpArr)
      
      // shift mat
      val shiftMat1 = new Mat(3, 3, CvType.CV_32F)
      tmpArr = Array[Float](
          1, 0, 0,
          0, 1, 0,
          0, 0, 1
      )
      shiftMat1.put(0, 0, tmpArr)
  
      tmpArr(2) = -dstPt(0)
      tmpArr(5) = -dstPt(1)
      
      shiftMat1.put(0, 0, tmpArr)  
      
      // rotate mat
      val rotateMat = new Mat(3, 3, CvType.CV_32F)
      tmpArr = Array[Float](
          1, 0, 0,
          0, 1, 0,
          0, 0, 1
      )
      rotateMat.put(0, 0, tmpArr)
  
      val angle = opt.rotateAngle / 180.0 * 3.14159265358979323846
      val cosA = Math.cos(angle).toFloat
      val sinA = Math.sin(angle).toFloat
      
      tmpArr(0) = cosA
      tmpArr(1) = sinA
      tmpArr(3) = -sinA
      tmpArr(4) = cosA
      
      rotateMat.put(0, 0, tmpArr)

      // shear mat
      val shearMat = new Mat(3, 3, CvType.CV_32F)
      tmpArr = Array[Float](
          1, 0, 0,
          0, 1, 0,
          0, 0, 1
      )
      shearMat.put(0, 0, tmpArr)
      
      tmpArr(1) = opt.shearFactor
      tmpArr(3) = opt.shearFactor
      
      shearMat.put(0, 0, tmpArr)

      // shift mat
      val shiftMat2 = new Mat(2, 3, CvType.CV_32F)
      tmpArr = Array[Float](
          1, 0, 0,
          0, 1, 0
      )
      shiftMat2.put(0, 0, tmpArr)
  
      tmpArr(2) = dstPt(0)
      tmpArr(5) = dstPt(1)
      
      shiftMat2.put(0, 0, tmpArr)    

      val out = new Mat()
      val tranMat = new Mat()
      
      Core.gemm(shiftMat2, shearMat, 1, new Mat(), 0, tranMat)
      Core.gemm(tranMat, rotateMat, 1, new Mat(), 0, tranMat)
      Core.gemm(tranMat, shiftMat1, 1, new Mat(), 0, tranMat)
      Core.gemm(tranMat, cropMat, 1, new Mat(), 0, tranMat)
      Imgproc.warpAffine(img, out, tranMat, new Size(outWh(0), outWh(1)))
      
      Highgui.imwrite(opt.outPath, out)
      val cropShow = new Imshow("croped image")
      cropShow.showImage(out)
      
    } catch {
      case ex: Exception => {
        println(ex.getMessage, ex)
        parser.printUsage(System.err)
        sys.exit(1)
      }
    }
  }

  class Argments {
    @Option(name = "--image", usage = "input image")
    var image: String = "../datas/timg.jpg"
    @Option(name = "--bbox", usage = "")
    var bbox: String = "175,30,250,270" // top_left_x,top_left_y,width,height
    @Option(name = "--rotate-angle", usage = "")
    var rotateAngle: Float = 20
    @Option(name = "--shear-factor", usage = "")
    var shearFactor: Float = 0.2f
    @Option(name = "--out-wh", usage = "")
    var outWH: String = "250,270" // output_width,output_height
    @Option(name = "--out-path", usage = "")
    var outPath: String = "../datas/cropped.jpg" // output_width,output_height
  }
}

File: AffineTransformationEx/Python/affiex.py
import cv2
import argparse
import math
import numpy as np

def parse_args():
  parser = argparse.ArgumentParser(description='')
  parser.add_argument('--image', type=str, default='../datas/timg.jpg', help='')
  parser.add_argument('--out-path', type=str, default='../datas/cropped.jpg', help='')
  parser.add_argument('--bbox', type=str, default='175,30,250,270', help='')
  parser.add_argument('--rotate-angle', type=float, default=10, help='')
  parser.add_argument('--out-wh', type=str, default='250,270', help='')
  parser.add_argument('--shear-factor', type=float, default=0.1, help='')
  return parser.parse_args()

if __name__ == '__main__':
  args = parse_args()

  img = cv2.imread(args.image, 1)

  bbox = [int(e) for e in args.bbox.split(',')]
  out_wh = [int(e) for e in args.out_wh.split(',')]

  # crop matrix, mapping the center of bbox to the center of croped image
  crop_mat = np.zeros((3, 3), np.float32)
  crop_mat[0][0] = 1
  crop_mat[1][1] = 1
  crop_mat[2][2] = 1

  bbox_center_x = bbox[0] + bbox[2] / 2
  bbox_center_y = bbox[1] + bbox[3] / 2

  src_pt = [bbox_center_x, bbox_center_y]
  dst_pt = [out_wh[0] / 2, out_wh[1] / 2]

  scale_x = float(out_wh[0]) / bbox[2]
  scale_y = float(out_wh[1]) / bbox[3]

  crop_mat[0][0] = scale_x
  crop_mat[1][1] = scale_y

  crop_mat[0][2] = dst_pt[0] - crop_mat[0][0] * src_pt[0]
  crop_mat[1][2] = dst_pt[1] - crop_mat[1][1] * src_pt[1]

  # shift matrix 
  shift_mat1 = np.zeros((3, 3), np.float32)
  shift_mat1[0][0] = 1
  shift_mat1[1][1] = 1
  shift_mat1[2][2] = 1

  shift_mat1[0][2] = -dst_pt[0]
  shift_mat1[1][2] = -dst_pt[1]

  # rotate matrix
  rotate_mat = np.zeros((3, 3), np.float32)
  rotate_mat[0][0] = 1
  rotate_mat[1][1] = 1
  rotate_mat[2][2] = 1

  angle = args.rotate_angle / 180.0 * 3.14159265358979323846
  cos_a = math.cos(angle)
  sin_a = math.sin(angle)

  rotate_mat[0][0] = cos_a
  rotate_mat[0][1] = sin_a
  rotate_mat[1][0] = -sin_a
  rotate_mat[1][1] = cos_a

  # shear matrix
  shear_mat = np.zeros((3, 3), np.float32)
  shear_mat[0][0] = 1
  shear_mat[1][1] = 1
  shear_mat[2][2] = 1

  shear_x = args.shear_factor

  shear_mat[0][1] = shear_x
  shear_mat[1][0] = shear_x

  # shift matrix 
  shift_mat2 = np.zeros((2, 3), np.float32)
  shift_mat2[0][0] = 1
  shift_mat2[1][1] = 1

  shift_mat2[0][2] = dst_pt[0]
  shift_mat2[1][2] = dst_pt[1]

  tran_mat = cv2.gemm(shift_mat2, shear_mat, 1, None, 0) 
  tran_mat = cv2.gemm(tran_mat, rotate_mat, 1, None, 0) 
  tran_mat = cv2.gemm(tran_mat, shift_mat1, 1, None, 0) 
  tran_mat = cv2.gemm(tran_mat, crop_mat, 1, None, 0) 

  out = cv2.warpAffine(img, tran_mat, (out_wh[0], out_wh[1]))

  cv2.imshow('original image', img)
  cv2.imshow('cropped image', out)
  cv2.imwrite(args.out_path, out)
  cv2.waitKey()

      

